devito/symbolics/manipulation.py

```diff
@@ -319,12 +319,83 @@ def reuse_if_untouched(expr, args, evaluate=False):
         return expr.func(*args, evaluate=evaluate)
 
 
-def evalmin(a, b):
+def evalrel(func=min, input=None, assumptions=[]):
     """
-    Simplify min(a, b) if possible.
+    Simplify evalmin if possible and return nested MIN/MAX expression. This function
+    takes advantage of python's built-in `min`, Sympy's `Min` and our in-house `MIN`
+    to reduce multiple candidates in relation's when possible.
+
+    Parameters
+    ----------
+    func : min or max, the operation to follow
+        Defaults to min
+    input : a list of the candidate symbols to be simplified
+        The candidates to be evaluated. Defaults to None.
+    assumptions : A list of assumptions formed as relationals between candidates,
+        assumed to be True. Defaults to None.
+
+    Examples
+    --------
+    Assuming no values are known for `a`, `b`, `c`, `d` but we know that `d<=a` and
+    `c>=b` we can safely drop `a` and `c` from the candidate list.
+
+    >>> from devito import Symbol
+    >>> a = Symbol('a')
+    >>> b = Symbol('b')
+    >>> c = Symbol('c')
+    >>> d = Symbol('d')
+    >>> evalrel(max, [a, b, c, d], [[Le(d, a), Ge(c, b)]])
+    MAX(a, c)
     """
+    sfunc = (Min if func is min else Max)  # Choose sympy's Min/Max
+    rfunc = (rmin if func is min else rmax)  # Choose built-in MIN/MAX
+
+    mapper = {}
+
+    processed = []
+    for asm in assumptions:
+        if (asm.__class__ in (Ge, Gt)) and asm.rhs.is_Add and asm.lhs.is_positive:
+            # If `c >= a + b, {a, b, c} >= 0` then add 'c>=a, b>=a'
+            if all(i.is_positive for i in asm.rhs.args):
+                processed.extend(Ge(asm.lhs, i) for i in asm.rhs.args)
+            # If `c >= a + b, a>= 0, b<=0` then add 'c>=b, c<=a'
+            elif len(asm.rhs.args) == 2:
+                processed.extend(Lt(asm.lhs, i) if i.is_positive else
+                                 Gt(asm.lhs, i) for i in asm.rhs.args)
+        elif (asm.__class__ in (Le, Lt)) and asm.lhs.is_Add and asm.rhs.is_positive:
+            # If `c >= a + b, {a, b, c} >= 0` then add 'c>=a, b>=a'
+            if all(i.is_positive for i in asm.lhs.args):
+                processed.extend(Le(i, asm.rhs) for i in asm.lhs.args)
+            elif len(asm.lhs.args) == 2:
+                processed.extend(Lt(i, asm.rhs) if i.is_positive else
+                                 Gt(i, asm.rhs) for i in asm.lhs.args)
```

Comment:
I find these two cases a bit weird since they seem to be exactly the same, i.e check "c>= a + b"  or "a+b <=c". If it's actually the case the `elif` isn't needed you just need one line to convert `Le/Lt` into `Ge/Gt` to only consider the single unique case.

Reply:
Fixed